src/main/java/org/spongepowered/api/command/AbstractCommandModule.java

```diff
@@ -0,0 +1,366 @@
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.command;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import org.slf4j.Logger;
+import org.spongepowered.api.Game;
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.command.annotation.Command;
+import org.spongepowered.api.command.annotation.PlainDescription;
+import org.spongepowered.api.command.args.CommandArgs;
+import org.spongepowered.api.command.args.parsing.InputTokenizer;
+import org.spongepowered.api.command.binding.Binder;
+import org.spongepowered.api.command.binding.BindingBuilder;
+import org.spongepowered.api.command.binding.BindingKey;
+import org.spongepowered.api.command.binding.ReadOnlyDelegateBinder;
+import org.spongepowered.api.command.binding.simple.SimpleBinder;
+import org.spongepowered.api.plugin.PluginContainer;
+import org.spongepowered.api.text.serializer.TextSerializer;
+import org.spongepowered.api.text.serializer.TextSerializers;
+import org.spongepowered.api.text.translation.TextFunction;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+import javax.annotation.Nullable;
+
+/**
+ * A command module contributes configuration information to commands, and
+ * registers commands with a {@link CommandManager}.
+ *
+ * <p>A command module is <b>not</b> re-usable.</p>
+ */
+public abstract class AbstractCommandModule {
+
+    private Binder binder = new SimpleBinder();
+    @SuppressWarnings("NullableProblems")
+    private PluginContainer plugin;
+    @SuppressWarnings("NullableProblems")
+    private CommandManager manager;
+    private InputTokenizer inputTokenizer = InputTokenizer.quotedStrings(false);
+    private TextSerializer textSerializer = TextSerializers.FORMATTING_CODE;
+    @Nullable private CommandThrowableProcessor throwableProcessor;
+    @Nullable private TextFunction textFunction;
+    private CommandAuthorizer authorizer = CommandAuthorizer.CALLABLE;
+    private boolean configured;
+    private boolean registered;
+
+    protected AbstractCommandModule() {
+    }
+
+    /**
+     * Contributes bindings and other configurations for this module to {@code binder}.
+     *
+     * @param plugin The plugin
+     * @param manager The command manager
+     */
+    @SuppressWarnings("ConstantConditions")
+    public final void configure(PluginContainer plugin, CommandManager manager) {
+        checkState(this.manager == null, "Re-entry is not allowed.");
+        this.plugin = checkNotNull(plugin, "plugin");
+        this.manager = checkNotNull(manager, "manager");
+        this.preConfigure();
+        this.configure();
+        // Create a delegating read-only binder
+        this.binder = new ReadOnlyDelegateBinder(this.binder);
+        this.configured = true;
+    }
+
+    private void preConfigure() {
```

Comment:
Guice modules don't work the same, nor is it really a good idea to use it since guice does things differently. feel free to look into it more if you want